# Taller — Data Cleaning 
**Dataset:** `empleados_empresa.csv`  
**Autor:** Robert Moreno  

> En este taller trabajaremos exclusivamente limpieza y exploración básica de datos.
> Usa buenas prácticas y comenta tu código cuando sea necesario.


## 1) Leer el dataset

Lee `employees_workshop.csv` asegurando que:
- `id_empleado` se lea como texto
- El DataFrame se guarde en una variable llamada `df`
- Elimine la columna usando `del df["nombre_columna"]` en el mismo bloque de la lectura

> **Sugerencia:** Solo usa los hints si no sabes cómo partir

<details>
<summary><b>💡 Hints</b></summary>

- Hay 2 columnas que comienzan con "Unnamed" que deben eliminarse
- Usa `.info()` después de leer para identificarlas

</details>

In [117]:
# Tu código aquí
import pandas as pd

#Actualizar línea
df = pd.read_csv("employees_workshop.csv", sep =";")

In [118]:
del df["Unnamed: 0.1"]
del df["Unnamed: 0"]
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 956 entries, 0 to 955
Data columns (total 10 columns):
 #   Column                Non-Null Count  Dtype
---  ------                --------------  -----
 0   id_empleado           956 non-null    int64
 1   sexo                  956 non-null    str  
 2   edad                  956 non-null    int64
 3   ciudad                956 non-null    str  
 4   departamento          956 non-null    str  
 5   salario               956 non-null    str  
 6   fecha_ingreso         956 non-null    str  
 7   evaluacion_desempeno  956 non-null    int64
 8   modalidad             956 non-null    str  
 9   estado_civil          956 non-null    str  
dtypes: int64(3), str(7)
memory usage: 74.8 KB


## 2) Verificar tipos de datos

- Identifica qué columnas deberían cambiar de tipo
- Corrige los tipos si es necesario (por ejemplo fechas)

<details>
<summary><b>💡 Hints</b></summary>

- **Salario**: Está como texto. Revisa el formato de los valores, ¿qué símbolo y separador decimal tiene? Necesitas limpiarlo antes de convertir a `float`
- **Fechas**: Tienen múltiples formatos mezclados. Usa `pd.to_datetime()` con el parámetro `format='mixed'`

</details>

In [119]:
# Tu código aquí
# Identificacion de columnas 
df.dtypes

id_empleado             int64
sexo                      str
edad                    int64
ciudad                    str
departamento              str
salario                   str
fecha_ingreso             str
evaluacion_desempeno    int64
modalidad                 str
estado_civil              str
dtype: object

In [120]:
# Corrección
# Salario
df["salario"] = df["salario"].str.replace("$", "").str.replace(",", ".").astype("float")
df["salario"]

0      3355.03
1      2393.25
2      2423.75
3      1058.84
4      1975.49
        ...   
951     904.53
952    3636.99
953    1511.15
954    5506.46
955    1524.95
Name: salario, Length: 956, dtype: float64

In [121]:
#fecha
df["fecha_ingreso"] = pd.to_datetime(df["fecha_ingreso"], format = "mixed")
df["fecha_ingreso"]

0     2024-09-05
1     2015-04-09
2     2015-07-15
3     2025-08-25
4     2020-08-12
         ...    
951   2023-12-14
952   2024-11-02
953   2016-03-09
954   2019-12-29
955   2015-05-06
Name: fecha_ingreso, Length: 956, dtype: datetime64[us]

## 3) Limpieza columnas de texto
Identifique problemas en las columnas de texto, y corríjalas. Guarde los cambios en el mismo dataframe

<details>
<summary><b>💡 Hints</b></summary>

- **Sexo**: Hay valores como 'female', 'm', 'M', 'F'. Estandariza a solo 'M' y 'F'
- **Ciudad**: Hay mayúsculas/minúsculas inconsistentes y abreviaturas ('GYE', 'UIO'). Estandariza y reemplaza las abreviaturas
- Usa `.unique()` para ver los valores únicos de cada columna de texto

</details>

In [122]:
# Tu código aquí
df["sexo"].unique()

<StringArray>
['female', 'm', 'M', 'F']
Length: 4, dtype: str

In [123]:
df["ciudad"].unique()

<StringArray>
['guayaquil', 'Guayaquil', 'GYE', 'quito', 'UIO']
Length: 5, dtype: str

In [124]:
df["sexo"] = df["sexo"].str.upper().str.replace("FEMALE", "F").str.replace("MALE", "M")
df["sexo"]

0      F
1      M
2      M
3      F
4      M
      ..
951    M
952    F
953    F
954    F
955    F
Name: sexo, Length: 956, dtype: str

In [125]:
df["ciudad"] = df["ciudad"].str.upper().str.replace("GUAYAQUIL", "GYE").str.replace("QUITO", "UIO")
df["ciudad"]

0      GYE
1      GYE
2      GYE
3      GYE
4      UIO
      ... 
951    UIO
952    GYE
953    UIO
954    GYE
955    GYE
Name: ciudad, Length: 956, dtype: str

## 4) Valores nulos

- ¿Existen valores nulos en el dataset?
- ¿En qué columnas?
- ¿Cuántos?

<details>
<summary><b>💡 Hints</b></summary>

- Usa `.isnull().sum()` para contar nulos por columna
- O `.info()` para ver el conteo de non-null

</details>

In [126]:
# Tu código aquí
df.isna().sum()

# No hay valores nulos

id_empleado             0
sexo                    0
edad                    0
ciudad                  0
departamento            0
salario                 0
fecha_ingreso           0
evaluacion_desempeno    0
modalidad               0
estado_civil            0
dtype: int64

## 5) Duplicados

- ¿Existen registros duplicados?
- Verifícalo.
- Si existen, elimínalos.

<details>
<summary><b>💡 Hints</b></summary>

- `.duplicated().sum()` te dice cuántos hay
- `.drop_duplicates()` para eliminarlos (usa `inplace=True` para modificar el df original)

</details>

In [127]:
# Tu código aquí
df.duplicated().sum()
df[df.duplicated()]

,id_empleado,sexo,edad,ciudad,departamento,salario,fecha_ingreso,evaluacion_desempeno,modalidad,estado_civil
137,75591,F,26,GYE,Tecnologia,2200.43,2023-10-26,4,remoto,Soltero
269,56999,F,40,GYE,Ventas,2105.42,2020-11-18,4,remoto,Casado
528,25853,M,38,UIO,Ventas,1537.29,2017-05-21,5,presencial,Soltero
673,18595,M,42,UIO,Tecnologia,2953.46,2019-08-11,3,remoto,Casado
795,90574,F,27,GYE,Ventas,1060.55,2024-04-11,3,hibrido,Soltero


In [147]:
df = df.drop_duplicates()

In [148]:
df.duplicated().sum()

np.int64(0)

## 6) Exploración básica

¿Cuál es la edad mínima, máxima, y edad promedio de nuestro dataset?

<details>
<summary><b>💡 Hints</b></summary>

- `.min()`, `.max()`, `.mean()` sobre la columna 'edad'
- O usa `.describe()` para ver todas las estadísticas

</details>

In [136]:
# Tu código aquí
df["edad"].min()

np.int64(21)

In [138]:
df["edad"].max()

np.int64(54)

In [139]:
df["edad"].mean()

np.float64(36.30543933054393)

## 7) Departamentos

- ¿Qué departamentos existen?
- Muestra la proporción (%) de empleados por departamento.

<details>
<summary><b>💡 Hints</b></summary>

- `.unique()` para ver qué departamentos hay
- `.value_counts()` para contar empleados
- `.value_counts(normalize=True) * 100` para obtener porcentajes

</details>

In [ ]:
# Tu código aquí
df["departamento"].unique()

<StringArray>
['Ventas', 'Tecnologia', 'Operaciones', 'RRHH', 'Marketing', 'Finanzas']
Length: 6, dtype: str

In [141]:
df["departamento"].value_counts()

departamento
Tecnologia     332
Operaciones    180
Ventas         173
Finanzas       120
Marketing       90
RRHH            61
Name: count, dtype: int64

In [144]:
df_depa = df["departamento"].value_counts(normalize=True)*100
df_depa.round(2).astype(str) + "%"

departamento
Tecnologia     34.73%
Operaciones    18.83%
Ventas          18.1%
Finanzas       12.55%
Marketing       9.41%
RRHH            6.38%
Name: proportion, dtype: str

## 8) Distribución por género

- ¿Cuántos hombres y mujeres hay?
- ¿Qué porcentaje representa cada grupo?

<details>
<summary><b>💡 Hints</b></summary>

- Si ya limpiaste la columna 'sexo', deberías tener solo 'M' y 'F'
- Usa `.value_counts()` y `.value_counts(normalize=True) * 100`

</details>

In [145]:
# Tu código aquí
df["sexo"].value_counts()

sexo
M    499
F    457
Name: count, dtype: int64

In [146]:
df_sexo = df["sexo"].value_counts(normalize=True)*100
df_sexo.round(2).astype(str) + "%"

sexo
M    52.2%
F    47.8%
Name: proportion, dtype: str

## 9) Comparación salarial

¿Qué departamento tiene mayor salario promedio: **Tecnología** o **Finanzas**?

Calcula el promedio de salario para ambos y concluye.

<details>
<summary><b>💡 Hints</b></summary>

- Filtra el dataframe por departamento: `df[df['departamento'] == 'Tecnologia']`
- Luego aplica `.mean()` sobre la columna 'salario'
- O usa `.groupby('departamento')['salario'].mean()` para ambos a la vez

</details>

In [149]:
# Tu código aquí
df[df["departamento"] == "Tecnologia"]["salario"].mean()

np.float64(3413.8350303030306)

In [150]:
df[df["departamento"] == "Finanzas"]["salario"].mean()

np.float64(3115.8551666666667)

## 10) Pregunta libre

* Con los conocimientos adquiridos en este momento, usando filtros y/o agregaciones proponga 1 pregunta de exploración básica, que considere al menos 2 filtros en el dataset.

<details>
<summary><b>💡 Hints</b></summary>

- Combina condiciones con `&` (AND) u `|` (OR)
- Ejemplo: `df[(df['edad'] > 30) & (df['modalidad'] == 'remoto')]`
- No olvides los paréntesis en cada condición

</details>

In [ ]:
# Tu código aquí


Pregunta:
* ¿Cuál es el salario máximo del área de RRHH en Quito?

## 11) BONUS
 (Punto extra válido en cualquier Actividad; Solo válido si se resuelve durante la clase).

La directora de RRHH necesita saber si existe alguna brecha salarial entre hombres y mujeres en la empresa. 

Cuentas con la información de todos los departamentos y todos los empleados para poder responder esa pregunta.

Realice el análisis solicitado, y concluya

<details>
<summary><b>💡 Hints</b></summary>

- Usa `.groupby('sexo')['salario'].mean()` para comparar salarios promedio
- Calcula la diferencia y el porcentaje de brecha

</details>

In [156]:
brecha = df.groupby('sexo')['salario'].mean()
brecha.round(2)

sexo
F    2676.97
M    2706.73
Name: salario, dtype: float64

In [159]:
diferencia = brecha.iloc[0] - brecha.iloc[1]
diferencia.round(2)

np.float64(-29.76)

Resultado: Hay una diferencia de $29.76 en el promedio de salarios entre hombres y mujeres

---
### Entregable

Sube tu notebook con:
- Código limpio
- Resultados visibles
